### PR TITLE
Rtsan/blocking 2 llvm pass

### PR DESCRIPTION
**Pull request by davidtrevelyan**

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 1]
```diff
@@ -45,6 +45,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
         insertCallBeforeInstruction(Fn, I, InsertFnName);
 }
 
+static PreservedAnalyses rtsanPreservedAnalyses() {
+  PreservedAnalyses PA;
+  PA.preserveSet<CFGAnalyses>();
+  return PA;
+}
+
+static void transformRealtimeUnsafeFunction(Function &F) {
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:54]: Nit (take it or leave it):

Transform is a little vague, could be more descriptive as to the actual transformation taking place

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 2]
```diff
@@ -45,6 +45,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
         insertCallBeforeInstruction(Fn, I, InsertFnName);
 }
 
+static PreservedAnalyses rtsanPreservedAnalyses() {
+  PreservedAnalyses PA;
+  PA.preserveSet<CFGAnalyses>();
+  return PA;
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:51]: I would get rid of this function and leave these explicit. I think the loop changing we will do will not preserve CFG, so I think it is OK to just do this explicitly in our main block

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 3]
```diff
@@ -45,6 +45,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
...
+}
+
+static void transformRealtimeUnsafeFunction(Function &F) {
+  IRBuilder<> Builder(&F.front().front());
+  Value *NameArg = Builder.CreateGlobalString(F.getName());
+
+  FunctionType *FuncType =
+      FunctionType::get(Type::getVoidTy(F.getContext()),
+                        {PointerType::getUnqual(F.getContext())}, false);
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:60]: we should call reviewer eyes to this code when we post this so we make sure it is correct (not saying it isn't just that I have no clue about this code)

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+define noundef i32 @main() #2 {
+  call void @blocking_function() #4
+  ret i32 0
+}
+
+attributes #0 = { mustprogress noinline sanitize_realtime_unsafe optnone ssp uwtable(sync) }
+
+; RealtimeSanitizer pass should insert __rtsan_expect_not_realtime at function entrypoint
+; CHECK-LABEL: @blocking_function()
+; CHECK-NEXT: call{{.*}}@__rtsan_expect_not_realtime({{ptr .*}})
```
Comment (cjappl) [inline llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:16]: nice one!

**Review by cjappl — APPROVED** [review]
Overall this is good, a few small nits but this is very similar to how I would have it.

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 2 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:51]: Hmm yeah - part of my refactoring in the main pass function was intended to have a sequence of:

```cpp
if (condition) {
    return transformXYZ(...);
}
else if (condition) {
    return transformABC(...);
}
```

... to express the intent that these transformations are mutually exclusive and return early. But now I'm thinking about it, it doesn't really need to be that way - and it's a bit early days to start those sorts of refactors. So yeah, I think I'll revert this as you suggest

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 1 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:54]: Thanks, agreed

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 2 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:51]: On second thoughts, we still need it in two places. I've updated the function name to be more explicit, though - you'll see this in my latest PR. When we no longer need it as part of the `sanitize_realtime` pass, we can inline it then. I think it's better to keep it in this case, don't you?

**Pull request closed without merging**